tests/utilities/test_parsing.py

```diff
@@ -0,0 +1,192 @@
+
+
+def test_lightning_hasattr(tmpdir):
+    """ Test that the lightning_hasattr works in all cases"""
+    model1, model2, model3, model4 = _get_test_cases()
+    assert lightning_hasattr(model1, 'learning_rate'), \
+        'lightning_hasattr failed to find namespace variable'
+    assert lightning_hasattr(model2, 'learning_rate'), \
+        'lightning_hasattr failed to find hparams namespace variable'
+    assert lightning_hasattr(model3, 'learning_rate'), \
+        'lightning_hasattr failed to find hparams dict variable'
+    assert not lightning_hasattr(model4, 'learning_rate'), \
+        'lightning_hasattr found variable when it should not'
+
+
+def test_lightning_getattr(tmpdir):
+    """ Test that the lightning_getattr works in all cases"""
+    models = _get_test_cases()
+    for i, m in enumerate(models[:3]):
+        value = lightning_getattr(m, 'learning_rate')
+        assert value == i, 'attribute not correctly extracted'
+
+
+def test_lightning_setattr(tmpdir):
+    """ Test that the lightning_setattr works in all cases"""
+    models = _get_test_cases()
+    for m in models[:3]:
+        lightning_setattr(m, 'learning_rate', 10)
+        assert lightning_getattr(m, 'learning_rate') == 10, \
+            'attribute not correctly set'
+
+
+def test_str_to_bool_or_str(tmpdir):
+    true_cases = ['y', 'yes', 't', 'true', 'on', '1']
+    false_cases = ['n', 'no', 'f', 'false', 'off', '0']
+    other_cases = ['yyeess', 'noooo', 'lightning']
+
+    for case in true_cases:
+        assert str_to_bool_or_str(case) is True
+
+    for case in false_cases:
+        assert str_to_bool_or_str(case) is False
+
+    for case in other_cases:
+        assert str_to_bool_or_str(case) == case
+
+
+def test_str_to_bool(tmpdir):
+    true_cases = ['y', 'yes', 't', 'true', 'on', '1']
+    false_cases = ['n', 'no', 'f', 'false', 'off', '0']
+    other_cases = ['yyeess', 'noooo', 'lightning']
+
+    for case in true_cases:
+        assert str_to_bool(case) is True
+
+    for case in false_cases:
+        assert str_to_bool(case) is False
+
+    for case in other_cases:
+        with pytest.raises(ValueError):
+            str_to_bool(case)
+
+
+def test_is_picklable(tmpdir):
+    # See the full list of picklable types at
+    # https://docs.python.org/3/library/pickle.html#pickle-picklable
+    class UnpicklableClass:
+        # Only classes defined at the top level of a module are picklable.
+        pass
+
+    def unpicklable_function():
+        # Only functions defined at the top level of a module are picklable.
+        pass
+
+    true_cases = [None, True, 123, "str", (123, "str"), max]
+    false_cases = [unpicklable_function, UnpicklableClass]
+
+    for case in true_cases:
+        assert is_picklable(case) is True
+
+    for case in false_cases:
+        assert is_picklable(case) is False
+
+
+def test_clean_namespace(tmpdir):
+    class UnpicklableClass:
+        pass
+
+    def unpicklable_function():
+        pass
```

Comment:
there are repeating, can we define them outside the func?

Reply:
These must be defined at non-top level of a module in order for them to be unpicklable, so if we define them outside the function, they become picklable and thus we cannot test `is_picklable` when inputs are unpicklable.

Reply:
@akihironitta You could use a lambda instead?

Reply:
@justusschock Yes, I think we can.